# Generación de textos con Deep Learning

Las redes neuronales recurrentes también pueden usarse como modelos generativos. Esto significa que, además de ser utilizados para modelos predictivos, pueden aprender las secuencias de un problema y luego generar secuencias plausibles completamente nuevas para el dominio del problema. Los modelos generativos son útiles no solo para estudiar qué tan bien un modelo ha aprendido un problema, sino para aprender más sobre el dominio del problema en sí mismo. En este proyecto, se verá cómo crear un modelo generativo de texto, carácter por carácter, utilizando las redes neuronales recurrentes de LSTM en Python con Keras. 

## **Descripción del problema: Generación de texto**
Muchos de los textos clásicos ya no están protegidos por derechos de autor. Esto significa que se pueden descargar todo el texto de estos libros de forma gratuita y utilizarlos en experimentos, como crear modelos generativos. En este caso vamos a utilizar un libro de la infancia como el conjunto de datos: Los tres cerditos.
Vamos a aprender las dependencias entre los caracteres y las probabilidades condicionales de los caracteres en las secuencias para que a su vez podamos generar secuencias de caracteres totalmente nuevas y originales. 

## **¿Qué utilidad tienen los modelos generativos?**
Estos modelos permiten en base a un conjunto de datos apreder y generar datos siguiendo secuencias aprendidas bajo los datos. Esto se utiliza mucho en **phising**. En base a mesajes preestablecidos, se generan emails, cuentas de usuarios, mensajes en aplicaciones de mensajería... etc.. para engañar al usuarios simulando que se encuentra en una conversión real.

## **Desarrollar una pequeña Red Neuronal LSTM Recurrente**
Empezaremos a preparar el conjunto de datos para el modelado. Nuestro libro de ejemplo no tiene encabezado o pie de pagina, pero en caso de que lo tenga deberemos eliminarlo.

Comenzaremos desarrollando una sencilla red LSTM para aprender secuencias de caracteres de Los tres cerditos. En este punto usaremos este modelo para generar nuevas secuencias de caracteres. Empecemos por importar las clases y funciones que pretendemos usar para entrenar a nuestro modelo.

In [3]:
import sys
import numpy
import keras
from keras.models import Sequential
from keras.layers import Dense
from keras.layers import Dropout
from keras.layers import LSTM
from keras.callbacks import ModelCheckpoint
from keras.utils import np_utils

A continuación tendremos que cargar el texto ASCII del libro en la memoria y convertir todos los caracteres a minúsculas para reducir el vocabulario que la red debe aprender.

In [4]:
#Cargamos el texto y lo pasamos a minuscula
filename = "los3.txt";
raw_text = open(filename).read()
raw_text = raw_text.lower()

Ahora que el libro está cargado, debemos preparar los datos para el modelado de la red neuronal. No podemos modelar los caracteres directamente, sino que debemos convertirlos en enteros. Podemos hacer esto fácilmente creando primero un conjunto de todos los caracteres distintos en el libro, y luego creando un mapa de cada carácter con un número entero único.

In [5]:
#Crear mapeo de caracteres únicos a enteros, y un mapeo inverso
chars = sorted(list(set(raw_text)))
char_to_int = dict((c, i) for i, c in enumerate(chars))

Por ejemplo, la lista de caracteres únicos en minúsculas ordenados en el libro es la siguiente:  
  
['\n', '\r', ' ', '!', '"', "'", '(', ')', '*', ',', '-', '.', ':', ';', '?', '[', ']',
'_', 'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p',
'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z', '\xbb', '\xbf', '\xef'].  
  
Podemos ver que puede haber algunos caracteres que podríamos eliminar para limpiar aún más el conjunto de datos reduciendo el vocabulario y mejorando el proceso de modelado. Ahora que el libro ha sido cargado y el mapeo preparado, podemos resumir el conjunto de datos.

In [6]:
n_chars = len(raw_text)
n_vocab = len(chars)
print("Total de caracteres:", n_chars)
print("Total vocales:", n_vocab)

Total de caracteres: 3139
Total vocales: 38


Podemos ver que el libro tiene algo menos de 3139 caracteres y que cuando se convierte a minúsculas sólo hay 38 caracteres distintos en el vocabulario para que la red aprenda. Mucho más que los 27 del alfabeto. Ahora tenemos que analizar los datos de formación para la red. Hay mucha flexibilidad en la forma que se decide dividir el texto y exponerlo a la red durante la formación. Aquí lo dividiremos en las siguientes secciones con una longitud fija de 100 caracteres, una longitud arbitraria. Podríamos fácilmente dividir los datos por frases y rellenar las secuencias más cortas y truncar las más largas.

Cada patrón de formación de la red se compone de 100 pasos de tiempo de un carácter (X) seguido de la salida de un carácter (y). Al crear estas secuencias, deslizamos esta ventana a lo largo de todo el libro, un carácter a la vez, lo que permite que cada carácter tenga la oportunidad de aprender de los 100 caracteres que lo precedieron (excepto los primeros 100 caracteres, por supuesto). Por ejemplo, si la longitud de la secuencia es 5 (por simplicidad) entonces los dos primeros patrones de entrenamiento serían los siguientes:

CHAPT -> E  
HAPTE -> R

A medida que dividimos el libro en estas secuencias, convertimos los caracteres en enteros usando nuestra tabla de búsqueda que preparamos anteriormente.

In [7]:
#Preparar el conjunto de datos de entrada para los pares de salida codificados como enteros
seq_length = 100
dataX = []
dataY = []
for i in range(0, n_chars - seq_length, 1):
  seq_in = raw_text[i:i + seq_length]
  seq_out = raw_text[i + seq_length]
  dataX.append([char_to_int[char] for char in seq_in])
  dataY.append(char_to_int[seq_out])
n_patterns = len(dataX)
print("Total patrones: ", n_patterns)

Total patrones:  3039


La ejecución del código hasta este punto nos muestra que cuando dividimos el conjunto de datos en datos de formación para que la red se entere de que tenemos algo menos de Total Patterns: 3039 patrones de formación. Esto tiene sentido ya que excluyendo los primeros 100 caracteres, tenemos un patrón de entrenamiento para predecir cada uno de los caracteres restantes.

  
Ahora que hemos preparado nuestros datos de entrenamiento, necesitamos transformarlos para que sean adecuados para su uso con Keras. Primero debemos transformar la lista de secuencias de entrada en la forma[muestras, pasos de tiempo, características] esperada por una red LSTM. A continuación necesitamos reescalar los números enteros al rango 0-a-1 para hacer que los patrones sean más fáciles de aprender por la red LSTM que utiliza la función de activación sigmoide por defecto.  
  
Finalmente, necesitamos convertir los patrones de salida (caracteres individuales convertidos en enteros) en una codificación en caliente. Esto es para que podamos configurar la red para predecir la probabilidad de cada uno de los 47 caracteres diferentes en el vocabulario (una representación más fácil) en lugar de tratar de forzarlo a predecir con precisión el siguiente carácter. Cada valor de y se convierte en un vector disperso con una longitud de 38, lleno de ceros excepto con un 1 en la columna para la letra (entero) que el patrón representa. Por ejemplo, cuando n (valor entero 31) es un valor codificado en caliente, se ve como como sigue:

[ 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.  
0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 1. 0. 0. 0. 0.  
0. 0. 0. 0. 0. 0. 0. 0.]  
  
Podemos implementar estos pasos como:

In [8]:
#Remodelar X para que sea [muestras, pasos de tiempo, características]
X = numpy.reshape(dataX, (n_patterns, seq_length, 1))
#Normalizacion
X = X / float(n_vocab)
#Codificacion en caliente con la variable de salida
y = np_utils.to_categorical(dataY)

Ahora podemos definir nuestro modelo LSTM. Aquí definimos una única capa LSTM oculta con 256 unidades de memoria. La red utiliza el abandono con una probabilidad del 20%. La capa de salida es una capa densa que utiliza la función de activación de softmax para generar una predicción de probabilidad para cada uno de los 38 caracteres entre 0 y 1. El problema es en realidad un problema de clasificación de un solo carácter con 38 clases y, como tal, se define como la optimización del pérdida de registro (entropía cruzada), utilizando el algoritmo de optimización de ADAM para la velocidad.

In [9]:
#Define el LSTM model
model = Sequential()
model.add(LSTM(256, input_shape=(X.shape[1], X.shape[2])))
model.add(Dropout(0.2))
model.add(Dense(y.shape[1], activation='softmax'))
model.compile(loss='categorical_crossentropy', optimizer='adam')

Instructions for updating:
Colocations handled automatically by placer.


No existe un conjunto de datos de prueba. Estamos modelando todo el conjunto de datos de entrenamiento para aprender la probabilidad de cada carácter en una secuencia. No nos interesa el modelo más preciso (precisión de clasificación) del conjunto de datos de formación. Este sería un modelo que predice cada carácter en el conjunto de datos de entrenamiento perfectamente. En su lugar, estamos interesados en una generalización del conjunto de datos que minimice la función de pérdida elegida. Buscamos un equilibrio entre la generalización y la adaptación, pero sin memorizar.
La red tarda en entrenarse (unos 35 segundos por epoch en mi pc usando la CPU). Dada la lentitud y debido a nuestros requisitos de optimización, utilizaremos el modelo de checkpointing para registrar todos los pesos de la red cada vez que se observe una mejora en las pérdidas en el fin de la epoch. Usaremos el mejor conjunto de pesos (la pérdida más baja) para instanciar nuestro sistema generativo en el siguiente punto.

In [10]:
#Se define el checkpoint
filepath="pesos-los3-30-{epoch:02d}-{loss:.4f}.hdf5";
checkpoint = ModelCheckpoint(filepath, monitor='loss', verbose=1, save_best_only=True, mode='min')
callbacks_list = [checkpoint]

Ahora podemos ajustar nuestro modelo a los datos. Aquí utilizamos 50 epoch y un tamaño de lote grande de 128 patrones.

In [11]:
#Ajuste del modelo
model.fit(X, y, epochs=50, batch_size=128, callbacks=callbacks_list)

Instructions for updating:
Use tf.cast instead.
Epoch 1/50
3039/3039 [==============================] - 8s 3ms/step - loss: 3.1921

Epoch 00001: loss improved from inf to 3.19206, saving model to pesos-los3-30-01-3.1921.hdf5
Epoch 2/50
3039/3039 [==============================] - 8s 3ms/step - loss: 2.9909

Epoch 00002: loss improved from 3.19206 to 2.99090, saving model to pesos-los3-30-02-2.9909.hdf5
Epoch 3/50
3039/3039 [==============================] - 9s 3ms/step - loss: 2.9841

Epoch 00003: loss improved from 2.99090 to 2.98413, saving model to pesos-los3-30-03-2.9841.hdf5
Epoch 4/50
3039/3039 [==============================] - 8s 2ms/step - loss: 2.9739

Epoch 00004: loss improved from 2.98413 to 2.97385, saving model to pesos-los3-30-04-2.9739.hdf5
Epoch 5/50
3039/3039 [==============================] - 8s 3ms/step - loss: 2.9695

Epoch 00005: loss improved from 2.97385 to 2.96953, saving model to pesos-los3-30-05-2.9695.hdf5
Epoch 6/50
3039/3039 [=============================

3039/3039 [==============================] - 13s 4ms/step - loss: 2.2626

Epoch 00047: loss improved from 2.32184 to 2.26258, saving model to pesos-los3-30-47-2.2626.hdf5
Epoch 48/50
3039/3039 [==============================] - 14s 5ms/step - loss: 2.2267

Epoch 00048: loss improved from 2.26258 to 2.22668, saving model to pesos-los3-30-48-2.2267.hdf5
Epoch 49/50
3039/3039 [==============================] - 14s 5ms/step - loss: 2.1812

Epoch 00049: loss improved from 2.22668 to 2.18122, saving model to pesos-los3-30-49-2.1812.hdf5
Epoch 50/50
3039/3039 [==============================] - 14s 5ms/step - loss: 2.1367

Epoch 00050: loss improved from 2.18122 to 2.13666, saving model to pesos-los3-30-50-2.1367.hdf5


Podéis ver resultados diferentes debido a la naturaleza estocástica del modelo, y porque la dificultad de la semilla aleatoria en los modelos LSTM en obtener resultados 100% reproducibles. Esto no es una preocupación para este modelo generativo. Después de ejecutar el ejemplo, deberíais tener un número de puntos de control de peso en el directorio local. Podríamos borrarlos todos excepto el que tenga el menor valor de pérdida. Por ejemplo, cuando ejecuté este ejemplo, debajo estaba el punto de control con la menor pérdida que logré.  
  
**pesos-los3-30-50-2.1798.hdf5**  
  
La pérdida de la red disminuyó casi todas las epoch y esperó que la red pueda dejar de entrenar por muchas más epoch. En el siguiente punto veremos el uso de este modelo para generar nuevas secuencias de texto.

# Generación de texto con una Red LSTM
La generación de texto utilizando la red LSTM es relativamente sencilla. En primer lugar, se cargan los datos y se define la red exactamente de la misma manera que hemos visto en el punto anterior, excepto que los pesos de la red se cargan desde un punto de control y la red no necesita ser entrenada.

In [12]:
#Carga de los pesos de la red
filename="pesos-los3-30-50-2.1367.hdf5"
model.load_weights(filename)
model.compile(loss='categorical_crossentropy', optimizer='adam')

Además, al preparar el mapeo de caracteres únicos a enteros, también debemos crear un mapeo inverso que podamos usar para convertir los enteros de nuevo en caracteres para que podamos entender las predicciones.

In [13]:
int_to_char = dict((i, c) for i, c in enumerate(chars))

Por último, tenemos que hacer predicciones. La forma más sencilla de usar el modelo LSTM de Keras para hacer predicciones es comenzar primero con una secuencia de semillas como entrada, generar el siguiente carácter y luego actualizar la secuencia de semillas para añadir el carácter generado al final y recortar el primer carácter. Este proceso se repite mientras queramos predecir nuevos caracteres (por ejemplo, una secuencia de 1.000 caracteres de longitud). Podemos elegir un patrón de entrada aleatorio como nuestra semilla y luego imprimir los caracteres generados a medida que los generamos.

In [14]:
#Toma una semilla aleatoria
start = numpy.random.randint(0, len(dataX)-1)
pattern = dataX[start]
print("Semilla:")
print("\"", ''.join([int_to_char[value] for value in pattern]), "\"")
#Genera los carácteres
for i in range(1000):
  x = numpy.reshape(pattern, (1, len(pattern), 1))
  x = x / float(n_vocab)
  prediction = model.predict(x, verbose=0)
  index = numpy.argmax(prediction)
  result = int_to_char[index]
  seq_in = [int_to_char[value] for value in pattern]
  sys.stdout.write(result)
  pattern.append(index)
  pattern = pattern[1:len(pattern)]
print("\nHecho.")

Semilla:
"  hacia el lago. los cerditos no lo volvieron a ver.  el mayor de ellos regañó a los otros dos por ha "
soi ae lasiia d  ln dn  no  no  no le casoita de la ee casiiaad  e ledee  e leddno  e madee  a ¡uuién teme al lobo feroz, al lobo  al lobo!  - ¡quién teme al lobo feroz, al lobo feroz! a  cnuién teme al lobo feroz, al lobo feroz! -  ¡uuién teme al lobo feroz, al lobo feroz! -  ¡uuién teme al lobo feroz, al lobo feroz! -  ¡uuién teme al lobo feroz, al lobo feroz! -  ¡uuién teme al lobo feroz, al lobo feroz! -  ¡uuién teme al lobo feroz, al lobo feroz! -  ¡uuién teme al lobo feroz, al lobo feroz! -  ¡uuién teme al lobo feroz, al lobo feroz! -  ¡uuién teme al lobo feroz, al lobo feroz! -  ¡uuién teme al lobo feroz, al lobo feroz! -  ¡uuién teme al lobo feroz, al lobo feroz! -  ¡uuién teme al lobo feroz, al lobo feroz! -  ¡uuién teme al lobo feroz, al lobo feroz! -  ¡uuién teme al lobo feroz, al lobo feroz! -  ¡uuién teme al lobo feroz, al lobo feroz! -  ¡uuién teme al lobo fer

Ejecutando este ejemplo primero se produce la semilla aleatoria seleccionada, luego cada carácter a medida que se genera. Por ejemplo, a continuación se muestran los resultados de una ejecución de este generador de texto. La semilla aleatoria fue:  
  
*Semilla: " plar tan fuerte como el viento de invierno. sopló y sopló, pero la casita de ladrillos era muy resis  "*  
  
El texto generado con la semilla aleatoria (limpiado para presentación) fue:  
  
*-¡uuién teme al lobo feroz, al lobo feroz! -  ¡uui ne me lo  ee ceeeeaa  -¡uui neme al lobo feroz, al lobo feroz! -  ¡uui ne me lo  ee ceeeeaa  -¡uuiéé teme al lobo feroz, al lobo feroz! a  ¡uui ne mabo  le meeee  --¡uuiéé teme al lobo feroz, al lobo feroz! -  ¡uui ne me lo  ee ceeeeaa  -¡uuiéé teme al lobo feroz, al lobo feroz! a  ¡uui ne mabo  le meeee  --¡uuiéé teme al lobo feroz, al lobo feroz! -  ¡uui ne me lo  ee ceeeeaa  -¡uuiéé teme al lobo feroz, al lobo feroz! a  ¡uui ne mabo  le meeee  --¡uuiéé teme al lobo feroz, al lobo feroz! -  ¡uui ne me lo  ee ceeeeaa  -¡uuiéé teme al lobo feroz, al lobo feroz! a  ¡uui ne mabo  le meeee  --¡uuiéé teme al lobo feroz, al lobo feroz! -  ¡uui ne me lo  ee ceeeeaa  -¡uuiéé teme al lobo feroz, al lobo feroz! a  ¡uui ne mabo  le meeee  --¡uuiéé teme al lobo feroz, al lobo feroz! -  ¡uui ne me lo  ee ceeeeaa  -¡uuiéé teme al lobo feroz, al lobo feroz! a  ¡uui ne mabo  le meeee  --¡uuiéé teme al lobo feroz, al lobo feroz! -  ¡uui ne me lo  e  
Hecho.*  
  
Podemos notar algunas observaciones sobre el texto generado.  
  
Por lo general, se ajusta al formato de línea observado en el texto original de menos de 80 caracteres antes de una nueva línea.
Algunas de las palabras en secuencia tienen sentido, pero muchas no lo tienen.  
El hecho de que este modelo del libro, basado en caracteres, produzca resultados como estos es muy impresionante. Da una idea de las capacidades de aprendizaje de las redes LSTM. Los resultados son no perfectos.  
  
En el siguiente punto analizaremos la mejora de la calidad de los resultados mediante el desarrollo de un una red LSTM mucho más grande.

# Red Neuronal Recurrente LSTM más grande
Ahora vamos ha hacer los mismo creando una red mucho más grande. Mantendremos el mismo número de unidades de memoria en 256, pero añadiremos una segunda capa.

In [15]:
#Define la LSTM model
model = Sequential()
model.add(LSTM(256, input_shape=(X.shape[1], X.shape[2]), return_sequences=True))
model.add(Dropout(0.2))
model.add(LSTM(256))
model.add(Dropout(0.2))
model.add(Dense(y.shape[1], activation='softmax'))
model.compile(loss='categorical_crossentropy', optimizer='adam')

También cambiaremos el nombre del peso de los puntos de control para que podamos determinar la diferencia entre los pesos de esta red y el anterior (agregando la palabra más grande en el nombre del archivo):

**filename="pesos-los3grande-303-30-2.3081.hdf5"**  
  
Por último, aumentaremos el número de epoch de formación de 50 a 150 y reduciremos el tamaño del lote de 128 a 64 para dar a la red más oportunidades de actualización y aprendizaje. El código completo sería (con todos los pasos vistos):  

In [16]:
#Grande LSTM Network para generar texto 3 cerditos
import sys
import numpy
from keras.models import Sequential
from keras.layers import Dense
from keras.layers import Dropout
from keras.layers import LSTM
from keras.callbacks import ModelCheckpoint
from keras.utils import np_utils
#Cargamos el texto y lo pasamos a minuscula
filename = "los3.txt"
raw_text = open(filename).read()
raw_text = raw_text.lower()
#Crear mapeo de caracteres únicos a enteros, y un mapeo inverso
chars = sorted(list(set(raw_text)))
char_to_int = dict((c, i) for i, c in enumerate(chars))
#Sumarizamos los datos cargados
n_chars = len(raw_text)
n_vocab = len(chars)
print("Total caracters: ", n_chars)
print("Total vocabulario: ", n_vocab)
#Preparar el conjunto de datos de entrada para los pares de salida codificados como enteros
seq_length = 100
dataX = []
dataY = []
for i in range(0, n_chars - seq_length, 1):
  seq_in = raw_text[i:i + seq_length]
  seq_out = raw_text[i + seq_length]
  dataX.append([char_to_int[char] for char in seq_in])
  dataY.append(char_to_int[seq_out])
n_patterns = len(dataX)
print("Total patrones: ", n_patterns)
#Remodelar X para que sea [muestras, pasos de tiempo, características]
X = numpy.reshape(dataX, (n_patterns, seq_length, 1))
#Normalizacion
X = X / float(n_vocab)
#Codificacion en caliente con la variable de salida
y = np_utils.to_categorical(dataY)
#Se define el LSTM modelo
model = Sequential()
model.add(LSTM(256, input_shape=(X.shape[1], X.shape[2]), return_sequences=True))
model.add(Dropout(0.2))
model.add(LSTM(256))
model.add(Dropout(0.2))
model.add(Dense(y.shape[1], activation='softmax'))
model.compile(loss='categorical_crossentropy', optimizer='adam')
#Se define el checkpoint
filepath="pesos-los3grandes-303-{epoch:02d}-{loss:.4f}.hdf5";
checkpoint = ModelCheckpoint(filepath, monitor='loss', verbose=1, save_best_only=True, mode='min')
callbacks_list = [checkpoint]
#Ajuste del modelo
model.fit(X, y, epochs=150, batch_size=64, callbacks=callbacks_list)

Total caracters:  3139
Total vocabulario:  38
Total patrones:  3039
Epoch 1/150
3039/3039 [==============================] - 46s 15ms/step - loss: 3.0710

Epoch 00001: loss improved from inf to 3.07103, saving model to pesos-los3grandes-303-01-3.0710.hdf5
Epoch 2/150
3039/3039 [==============================] - 65s 21ms/step - loss: 2.9907

Epoch 00002: loss improved from 3.07103 to 2.99070, saving model to pesos-los3grandes-303-02-2.9907.hdf5
Epoch 3/150
 832/3039 [=======>......................] - ETA: 48s - loss: 3.0061

KeyboardInterrupt: 

La ejecución de este ejemplo lleva algún tiempo, dependiendo de tu ordenador. Después de ejecutar este ejemplo, se puede lograr una pérdida de aproximadamente 0.018. Por ejemplo, el mejor resultado que obtuve al ejecutar este modelo se almacenó en un archivo de punto de control con el nombre:

In [17]:
filename="pesos-los3grandes-303-150-0.0105.hdf5"

Como en el punto anterior, podemos utilizar este mejor modelo de la ejecución para generar texto. El único cambio que necesitamos hacer en el script de generación de texto del punto anterior está en la especificación de la topología de la red y desde qué archivo sembrar los pesos de la red. El código completo sería (con todos los pasos vistos):

In [18]:
#Carga de la red LSTM grande para generar texto
import sys
import numpy
from keras.models import Sequential
from keras.layers import Dense
from keras.layers import Dropout
from keras.layers import LSTM
from keras.utils import np_utils
#Cargamos el texto y lo pasamos a minuscula
filename = "los3.txt"
raw_text = open(filename).read()
raw_text = raw_text.lower()
#Crear mapeo de caracteres únicos a enteros, y un mapeo inverso
chars = sorted(list(set(raw_text)))
char_to_int = dict((c, i) for i, c in enumerate(chars))
int_to_char = dict((i, c) for i, c in enumerate(chars))
#Sumarizamos los datos cargados
n_chars = len(raw_text)
n_vocab = len(chars)
print("Total Characters: ", n_chars)
print("Total Vocab: ", n_vocab)
#Preparar el conjunto de datos de entrada para los pares de salida codificados como enteros
seq_length = 100
dataX = []
dataY = []
for i in range(0, n_chars - seq_length, 1):
    seq_in = raw_text[i:i + seq_length]
    seq_out = raw_text[i + seq_length]
    dataX.append([char_to_int[char] for char in seq_in])
    dataY.append(char_to_int[seq_out])
    n_patterns = len(dataX)

print("Total Patterns: ", n_patterns)
#Remodelar X para que sea [muestras, pasos de tiempo, características]
X = numpy.reshape(dataX, (n_patterns, seq_length, 1))
#Normalizacion
X = X / float(n_vocab)
#Codificacion en caliente de la variable de salida
y = np_utils.to_categorical(dataY)
#Define la LSTM model
model = Sequential()
model.add(LSTM(256, input_shape=(X.shape[1], X.shape[2]), return_sequences=True))
model.add(Dropout(0.2))
model.add(LSTM(256))
model.add(Dropout(0.2))
model.add(Dense(y.shape[1], activation='softmax'))
model.compile(loss='categorical_crossentropy', optimizer='adam')
#Carga de los pesos de la red
filename="pesos-los3grandes-303-150-0.0105.hdf5"
model.load_weights(filename)
model.compile(loss='categorical_crossentropy', optimizer='adam')
#Toma una semilla aleatoria
start = numpy.random.randint(0, len(dataX)-1)
pattern = dataX[start]
print("Semilla:")
print("\"", ''.join([int_to_char[value] for value in pattern]), "\"")
#Genera los carácteres
for i in range(1000):
  x = numpy.reshape(pattern, (1, len(pattern), 1))
  x = x / float(n_vocab)
  prediction = model.predict(x, verbose=0)
  index = numpy.argmax(prediction)
  result = int_to_char[index]
  seq_in = [int_to_char[value] for value in pattern]
  sys.stdout.write(result)
  pattern.append(index)
  pattern = pattern[1:len(pattern)]
print("\nHecho.")

Total Characters:  3139
Total Vocab:  38
Total Patterns:  3039
Semilla:
" no, no te vamos a abrir.  - pues si no me abrís... ¡soplaré y soplaré y la casita derribaré!  y se p "
uso a soplar tan fuerte como el viento de invierno. sopló y sopló, pero la casita de ladrillos era muy resistente y no conseguía derribarla. decidió trepar por la pared y entrar por la chimenea. se deslizó hacia abajo... y cayó en el caldero donde el cerdito mayor estaba hirviendo sopa de nabos. escaldado y con el estómago vacío salió huyendo hacia el lago. los cerditos no lo volvieron a ver.  el mayor de ellos regañó a los otros dos por haber sido tan perezosos y poner en peligro sus propias vidas, y si algún día vais por el bosque y veis tres cerdos, sabréis que son los tres cerditos porque les gusta cantar:  - ¡quién teme al lobo feroz, al lobo, al lobo!  - ¡quién teme al lobo feroz, al lobo feroz! - cantaban desde dentro los cerditos. el lobo estaba realmente enfadado y hambriento, y ahora deseaba comerse a 

Un ejemplo de ejecución de este script de generación de texto produce la salida a continuación. El texto semilla elegido al azar fue:  
  
Semilla:
*" l más pequeño-, la paja es blanda y se puede sujetar con facilidad. terminaré muy pronto y podré ir  "*  
  
El texto generado con la semilla (limpiado para presentación) fue:
  
*a jugar. el hermano mediano decidió que su casa sería de madera:  - puedo encontrar un montón de madera por los alrededores - explicó 
a sus hermanos, - construiré mi casa en un santiamén con todos estos troncos y me iré también a jugar.  cuando las tres casitas estuvieron terminadas, los cerditos cantaban y bailaban en la puerta, felices por haber acabado con el problema:  -¡quién teme al lobo feroz, al lobo, al lobo!  - ¡quién teme al lobo feroz, al lobo feroz! - cantaban desde dentro los cerditos. el lobo estaba realmente enfadado y hambriento, y ahora deseaba comerse a los tres cerditos más que nunca, y frente a la puerta dijo:  - ¡cerditos, abridme la puerta! - no, no, no, no te vamos a abrir.  - pues si no me abrís... ¡soplaré y soplaré y la casita derribaré!  y se puso a soplar tan fuerte como el viento de invierno. sopló y sopló, pero la casita de ladrillos era muy resistente y no conseguía derribarla. decidió trepar por la pared y entrar por la chimenea. se deslizó hacia abajo  
Hecho.*  
  
Podemos ver que generalmente hay menos errores de ortografía y el texto parece más realista. Estos son mejores resultados, pero hay todavía se podría mejorar mucho más.

# Ideas de extensión para mejorar el modelo
A continuación se muestran ideas que se pueden investigar para mejorar aún más el modelo:

- Predecir menos de 1.000 caracteres como salida para una semilla dada.
- Eliminar toda la puntuación del texto fuente y, por lo tanto, del vocabulario de los modelos.
- Prueba con una codificación en caliente para las secuencias de entrada.
- Aumentar el número de epoch de entrenamiento a 100 o cientos.
- Añadir más unidades de memoria a las capas y/o más capas.
- Experimentar con los factores de escala (temperatura) al interpretar las probabilidades de predicción.

